lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5431,96 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Type check the function subexpression to resolve a type for it if
     // possible.
     fnExpr = typeCheckChildIndependently(callExpr->getFn());
-    if (!fnExpr)
-      return true;
+    if (!fnExpr) {
+      auto hadError = [&]() {
+        return CS.TC.Diags.hadAnyError();
+      };
+      auto UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn());
+      if (!UDE)
+        return hadError();
+
+      auto baseExpr = UDE->getBase();
+      // Offer fix-it for using `subscript` instead of the operator.
+      // FIXME: This doesn't check generic parameters
+      if (baseExpr && UDE->getName().getBaseName() == "subscript") {
+        auto baseType = CS.getType(baseExpr);
+        // Look up subscript declarations and return if there are none.
+        auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                      DeclName(DeclBaseName::createSubscript()));
+        if (lookup.empty())
+          return hadError();
+
+        auto argExpr = typeCheckChildIndependently(callExpr->getArg(), Type(),
+                                                   CTP_CallArgument);
+        if (!argExpr)
+          return hadError();
+
+        SmallVector<Identifier, 2> scratch;
+        ArrayRef<Identifier> argLabels = callExpr->getArgumentLabels(scratch);
+        SmallVector<OverloadChoice, 2> choices;
+        auto hasTrailClosure = callArgHasTrailingClosure(argExpr);
+
+        for (auto candidate : lookup)
+          choices.push_back(OverloadChoice(baseType, candidate.getValueDecl(),
+                                           UDE->getFunctionRefKind()));
+
+        CalleeCandidateInfo candidates(baseType, choices,
+                                       hasTrailClosure, CS, true);
+
+        auto argType = CS.getType(argExpr);
+        auto params = swift::decomposeArgType(argType, argLabels);
+
+        using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+
+        candidates.filterList([&](UncurriedCandidate candidate) -> Closeness {
+          auto candFuncType = candidate.getUncurriedFunctionType();
```

Comment:
In general, this all feels pretty close to `CalleeCandidateInfo::evaluateCloseness`.

Reply:
If I could use `evaluateCloseness`, I would. But it doesn't account for subtypes, conformances, I'm not sure it will treat tuples as equal regardless of the labeling, it can't ignore labels on *arguments* as far as I see, and neither it does any generic stuff without diagnosing.